pystiche_papers/sanakoyeu_et_al_2018/_data.py

```diff
@@ -0,0 +1,218 @@
+from os import path
+from typing import Any, Dict, Optional, Sized, Tuple, Union, cast
+
+import torch
+from torch import nn
+from torch.utils.data import DataLoader, Dataset, Sampler
+from torch.utils.data.sampler import RandomSampler
+from torchvision.datasets.utils import download_and_extract_archive
+
+from pystiche.data import ImageFolderDataset
+from pystiche.image import transforms
+from pystiche.image.transforms import functional as F
+from pystiche.image.utils import extract_image_size
+from pystiche.misc import verify_str_arg
+
+from ..utils import OptionalGrayscaleToFakegrayscale
+
+__all__ = [
+    "ClampSize",
+    "image_transform",
+    "WikiArt",
+    "style_dataset",
+    "content_dataset",
+    "batch_sampler",
+    "image_loader",
+]
+
+
+class ClampSize(transforms.Transform):
+    # https://github.com/pmeier/adaptive-style-transfer/blob/07a3b3fcb2eeed2bf9a22a9de59c0aea7de44181/prepare_dataset.py#L49-L68
+    def __init__(
+        self,
+        min_edge_size: int = 800,
+        max_edge_size: int = 1800,
+        interpolation_mode: str = "bilinear",
+    ):
+        super().__init__()
+
+        if max_edge_size < min_edge_size:
+            raise ValueError(
+                f"max_edge_size cannot be smaller than min_edge_size: "
+                f"{max_edge_size} < {min_edge_size}"
+            )
+
+        self.max_edge_size = max_edge_size
+        self.min_edge_size = min_edge_size
+        self.interpolation_mode = interpolation_mode
+
+    def forward(self, image: torch.Tensor) -> torch.Tensor:
+        short_edge_size, long_edge_size = sorted(extract_image_size(image))
+
+        if (
+            short_edge_size >= self.min_edge_size
+            and long_edge_size <= self.max_edge_size
+        ):
+            return image
+
+        size: Union[int, Tuple[int, int]]
+        if long_edge_size > self.max_edge_size:
+            size = self.max_edge_size
+            edge = "long"
+        else:  # short_edge_size < self.min_edge_size
+            size = (
+                self.min_edge_size
+                if short_edge_size / self.min_edge_size > 0.25
+                else (self.min_edge_size, self.min_edge_size)
+            )
+            edge = "short"
+
+        return cast(
+            torch.Tensor,
+            F.resize(
+                image, size, edge=edge, interpolation_mode=self.interpolation_mode
+            ),
+        )
+
+    def _properties(self) -> Dict[str, Any]:
+        dct = super()._properties()
+        dct["min_edge_size"] = self.min_edge_size
+        dct["max_edge_size"] = self.max_edge_size
+        if self.interpolation_mode != "bilinear":
+            dct["interpolation_mode"] = self.interpolation_mode
+        return dct
+
+
+def image_transform(edge_size: int = 768,) -> transforms.ComposedTransform:
+    # TODO: Add image augmentator with random affine transform hsv transfomration and
```

Comment:
Could you list all needed transformations with their respective parameters? I'll handle it from there.